From 0716b9d845ac12eb4427a8b64d34a0166fc99703 Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Thu, 18 Jul 2024 11:31:06 +0530
Subject: [PATCH] wifi: mac80211: Avoid removing other non-tx BSS when link
 remove is scheduled

In current mac80211 implementation, whenever we have stop_ap for any
MBSSID non-tx BSS, we will remove the other non-tx BSS which are part of
same tx sdata. But, with the support for ML Link reconfigure removal,
any one can be removed individually.

Thus to add support for link reconfigure removal, avoid removing the
other non-tx BSS which are part of same tx sdata.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 net/mac80211/cfg.c | 1 -
 1 file changed, 1 deletion(-)

--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1764,6 +1764,7 @@ static int ieee80211_stop_ap(struct wiph
 	struct ieee80211_link_data *link =
 		sdata_dereference(sdata->link[link_id], sdata);
 	struct ieee80211_bss_conf *link_conf;
+	struct wireless_dev *wdev = dev->ieee80211_ptr;
 	LIST_HEAD(keys);
 
 	if (WARN_ON(!link))
@@ -1845,7 +1846,12 @@ static int ieee80211_stop_ap(struct wiph
 		ieee80211_free_key_list(local, &keys);
 	}
 
-	ieee80211_stop_mbssid(sdata, link_id);
+	/* With dynamic link removal support, any non-tx bss can be removed
+	 * individually. Hence, call ieee80211_stop_mbssid when the wdev is
+	 * removed through any other user application(s).
+	 */
+	if (wdev->is_netdev_going_down)
+		ieee80211_stop_mbssid(sdata, link_id);
 	link_conf->enable_beacon = false;
 	link_conf->mbssid_tx_vif = NULL;
 	link_conf->mbssid_tx_vif_linkid = -1;
